### DISCUSSION TITLE
Trashbin versioning

### DISCUSSION THREAD

**Discussion by acmpo6ou** [gsantner/markor#1518] [Ideas]
Hi Greg!
First of all, I really like Markor! All of its features are brilliant and very convenient, so thank you for creating it!

I have a suggestion, however:
The Trash feature: when I delete my notes they're gone, forever. But what if later I realize that I shouldn't have deleted them. There is no way to restore the deleted notes.
It would be cool if when you delete notes, they are just moved to the special Trash folder. And if you need to restore them you just go to that folder, select needed notes and chose the "Restore" option. 

Now, I could've just created this "Trash" folder myself and when I need to delete the notes, instead of pressing the delete button, I could've move the notes to the "Trash" folder. But the problem is that I have to move the notes manually. Also it would be better if when I restore them, the notes are moved to the folder from which I deleted them. 
For example, if I delete a **/test_folder/example/note.md** note, when I restore it – it's moved back to it's parent folder (i.e. **/test_folder/example/**). But it will be hard to do all this manually.

**Comment by gsantner** [comment]
> Hi Greg!

Hello! :D

>  The Trash feature: when I delete my notes they're gone, forever. But what if later I realize that I shouldn't have deleted them. There is no way to restore the deleted notes.

My opinion is: 
1) Deleting should be atomic and final - no leftovers, from the used application (what other software does in terms of backups/versions is not that application's beer). 
2)  I'm responsible for my own actions. When I delete something unintentionally, I didn't think well enough about it if it's still relevant - AND I didn't properly think about backup in case I really cared about it. So regardless the application you use, the device you use - if it's important think about making backups. Devices get destroyed, devices get lost.
3) Rather terminology - but I think the better way to call it overall is "Archiving" files

**Reply by acmpo6ou** [reply]
> I'm responsible for my own actions. When I delete something unintentionally, I didn't think well enough about it if it's still relevant

Well, you see:
We are _humans_, _we make mistakes_. I may delete something completely unintentionally, just accidentally pressing the delete button etc. Basically, when something gets deleted accidentally, _I didn't think at all_. But I lost my data. That's why _not me_, _the machine_ should think.

Anyway, you know better than me I suppose.
Regards!

**Reply by gsantner** [reply]
I'm not blaming anybody for his loss - thats what everybody has todo self on a wrong decision. 

I see "deletion" as something that should be final with no leftovers. If you want something that makes it not final, I call it Archiving. And that's what you can do with copying files to some other location, or more generally - making backups. Currently there is no specific "archiving" button though  in Markor (only a copy one) that would do such things automatically for you.

**Comment by aubreyz** [comment]
I agree there should not be internal duplication. After all if OP lost his phone (or the memory became corrupted) every file would be lost (including any trashbin). Archiving of any deleted files should take place via whatever backup mechanism OP has chosen to use. 

For example I mirror all my files to google drive every few hours with foldersync and also with rsync, and each of these retains backups of every file change.

**Comment by acmpo6ou** [comment]
I never maintained an open source project, so you are much more experienced than me. Markor is great: private, no bloatware etc. So it seems that you know what you are doing.

That's why I wrote:
> you know better than me I suppose



It would be cool to have some plugin system though, then I could write a small plugin for myself that implements features I need. Too bad adding the plugin system to Markor would, probably, be really hard.

Thanks for really good responsiveness!
We can close this discussion now, unless you have something to add.

**Reply by gsantner** [reply]
> I never maintained an open source project, so you are much more experienced than me. Markor is great: private, no bloatware etc. So it seems that you know what you are doing.

Thanks, great to hear when people enjoy it 😄 .

>  We can close this discussion now, unless you have something to add.

Unlike Issues, discussions can't be really "closed" as I see :D 

> plugin system

Well, Google is on the way to locking down literally everything and all. Like not even Markor itself being given the permission in future to access storage the way it used to. Now think of extensions/addons/plugins which have to fight these kind of games as well - they die out soon after the first release.  
My thought is - if somebody wants to help and improve A project, the best they can do is to contribute to THE PROJECT itself 😃

**Reply by acmpo6ou** [reply]
> Google is on the way to locking down literally everything and all. Like not even Markor itself being given the permission in future to access storage the way it used to

Oh, so you're having issues porting Markor to Android 12? I haven't had a need to see the changes to sdk 31 yet, so idk what  problems Google introduced this time.

> if somebody wants to help and improve A project, the best they can do is to contribute to THE PROJECT itself

But some changes aren't good for the project as a whole, they are good for only a small subset of people or one person even.

**Reply by gsantner** [reply]
nono, I wasted hours of time to get it [SDK31] working properly. It works all here, can install the app and it works as is.  
But google disagrees that a file manager & document management app is eligable for direct file access.

See <https://github.com/gsantner/markor/pull/1521>

![grafik](https://user-images.githubusercontent.com/6735650/145190103-039db1d6-66e0-497a-92b3-5157f1e82c20.png)

**Reply by acmpo6ou** [reply]
Ah, okay, I understand now.
